torch/csrc/autograd/variable.cpp

```diff
@@ -368,4 +373,56 @@ unsigned VariableHooks::_register_hook(const Tensor& self, std::function<Tensor(
   return idx;
 }
 
+TORCH_API void handle_view_on_rebase(DifferentiableViewMeta* diff_view_meta, bool indirect) {
+  /// See NOTE [ View + Inplace detection ] for justification of the logic below
+  if (diff_view_meta->allow_rebase_history != OnRebase::ALLOW_REBASE) {
+    auto grad_fn = diff_view_meta->grad_fn_.get();
+    std::string msg;
+    std::string modified_obj;
+    // Create the header for the error message.
+    if (indirect) {
+      modified_obj = "its base or another view of its base has been";
+    } else {
+      modified_obj = "is being";
+    }
+    if (grad_fn) {
+      msg = c10::str("Output ", diff_view_meta->output_nr_, " of ", grad_fn->name(), " is a view and ",
+                     modified_obj, " modified inplace.");
+    } else {
+      msg = c10::str("A view was created in no_grad mode and ", modified_obj, " modified inplace with grad mode enable.");
+    }
+
+    if (diff_view_meta->allow_rebase_history == OnRebase::ERROR_REBASE) {
+      TORCH_CHECK(false, msg, " This view is the output of a function that returns multiple views. Such functions do not"
+                         " allow the output views to be modified inplace.");
+    } else {
+      if (diff_view_meta->allow_rebase_history == OnRebase::WARN_REBASE) {
+        TORCH_INTERNAL_ASSERT(!grad_fn);
+        msg = c10::str(msg, " Given that this use case is ambiguous and error-prone, it is deprecated and will be forbidden"
+                       "  starting 1.6. You can clarify your code and remove this warning by moving both the view and the"
+                       " inplace either both inside the no_grad block (if you don't want the inplace to be tracked) or both"
+                       " outside (if you want the inplace to be tracked).");
+      } else if (diff_view_meta->allow_rebase_history == OnRebase::WARN_REBASE_FUNCTION) {
+        msg = c10::str(msg, " This view was created inside a custom Function (or because an input was returned as-is) and the"
+                       " autograd logic to handle view+inplace would override the custom backward associated with the custom"
+                       " Function, leading to incorrect gradients. This behavior is deprecated and will be forbidden starting"
+                       " version 1.6. You can remove this warning by cloning the output of the custom Function.");
+      } else {
+        TORCH_INTERNAL_ASSERT(false, "Invalid OnRebase state");
+      }
+
+      if (!indirect && !grad_fn) {
+        // This view is (wrongly) detected as a leaf that requires grad and would raise the surprising: "a leaf Variable that
+        // requires grad is being used in an in-place operation." after the warning. So we make the warning an error directly.
+        TORCH_CHECK(false, msg);
+      } else {
+        TORCH_WARN(msg);
+      }
+    }
+
+    // We warn only once per view
+    diff_view_meta->allow_rebase_history = OnRebase::ALLOW_REBASE;
```

Comment:
Err... this isn't thread safe, right?

Reply:
(Or, this function is only ever called from inplace operations?)

Reply:
This is only ever called for inplace operations.
Also I did not added strict check here as the race would only make it warn twice if you modify the same Tensor inplace from two threads at the same time (sounds quite dangerous already) but can't lead to anything else.
I'll update the comment.